File: day15/main.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

type (
	Item struct {
		value Point
		g, f  int
		index int
	}

	PriorityQueue struct {
		q []*Item
		s map[Point]int
	}

	Point struct {
		x int
		y int
	}

	OpenSet struct {
		q *PriorityQueue
	}

	ClosedSet map[Point]struct{}
)

func NewPriorityQueue() *PriorityQueue {
	return &PriorityQueue{
		s: map[Point]int{},
	}
}

func (q PriorityQueue) Len() int { return len(q.q) }
func (q PriorityQueue) Less(i, j int) bool {
	return q.q[i].f < q.q[j].f
}
func (q PriorityQueue) Swap(i, j int) {
	q.q[i], q.q[j] = q.q[j], q.q[i]
	q.q[i].index = i
	q.q[j].index = j
	q.s[q.q[i].value] = i
	q.s[q.q[j].value] = j
}
func (q *PriorityQueue) Push(x interface{}) {
	n := len(q.q)
	item := x.(*Item)
	item.index = n
	q.q = append(q.q, item)
	q.s[item.value] = n
}
func (q *PriorityQueue) Pop() interface{} {
	n := len(q.q)
	item := q.q[n-1]
	q.q[n-1] = nil  // avoid memory leak
	item.index = -1 // for safety
	q.q = q.q[:n-1]
	delete(q.s, item.value)
	return item
}
func (q *PriorityQueue) Update(value Point, g, f int) bool {
	idx, ok := q.s[value]
	if !ok {
		return false
	}
	item := q.q[idx]
	item.g = g
	item.f = f
	heap.Fix(q, item.index)
	return true
}

func NewOpenSet() *OpenSet {
	return &OpenSet{
		q: NewPriorityQueue(),
	}
}

func (s *OpenSet) Empty() bool {
	return s.q.Len() == 0
}

func (s *OpenSet) Has(val Point) bool {
	_, ok := s.q.s[val]
	return ok
}

func (s *OpenSet) Get(val Point) (*Item, bool) {
	idx, ok := s.q.s[val]
	if !ok {
		return nil, false
	}
	return s.q.q[idx], true
}

func (s *OpenSet) Push(value Point, g, f int) {
	heap.Push(s.q, &Item{
		value: value,
		g:     g,
		f:     f,
	})
}

func (s *OpenSet) Pop() (Point, int, int) {
	item := heap.Pop(s.q).(*Item)
	return item.value, item.g, item.f
}

func (s *OpenSet) Update(value Point, g, f int) bool {
	return s.q.Update(value, g, f)
}

func NewClosedSet() ClosedSet {
	return ClosedSet{}
}

func (cs ClosedSet) Has(val Point) bool {
	_, ok := cs[val]
	return ok
}

func (cs ClosedSet) Push(val Point) {
	cs[val] = struct{}{}
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func manhattan(a, b Point) int {
	return abs(a.x-b.x) + abs(a.y-b.y)
}

func neighbors(grid [][]int, p Point) []Point {
	return []Point{
		{x: p.x - 1, y: p.y},
		{x: p.x, y: p.y - 1},
		{x: p.x + 1, y: p.y},
		{x: p.x, y: p.y + 1},
	}
}

func inBounds(p Point, w, h int) bool {
	return p.x >= 0 && p.x < w && p.y >= 0 && p.y < h
}

func getVal(grid [][]int, p Point, w, h int) int {
	return (grid[p.y%h][p.x%w]+p.x/w+p.y/h-1)%9 + 1
}

func pathfind(grid [][]int, vw, vh int, start, end Point) int {
	w := len(grid[0])
	h := len(grid)
	openSet := NewOpenSet()
	openSet.Push(start, 0, manhattan(start, end))
	closedSet := NewClosedSet()
	for !openSet.Empty() {
		cur, curg, _ := openSet.Pop()
		closedSet.Push(cur)
		if cur == end {
			return curg
		}
		for _, i := range neighbors(grid, cur) {
			if !inBounds(i, vw, vh) || closedSet.Has(i) {
				continue
			}
			g := curg + getVal(grid, i, w, h)
			f := g + manhattan(i, end)
			if v, ok := openSet.Get(i); ok {
				if g < v.g {
					openSet.Update(i, g, f)
				}
				continue
			}
			openSet.Push(i, g, f)
		}
	}
	return -1
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var grid [][]int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := []byte(scanner.Text())
		row := make([]int, 0, len(line))
		for _, i := range line {
			row = append(row, int(i)-'0')
		}
		grid = append(grid, row)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Part 1:", pathfind(grid, len(grid[0]), len(grid), Point{0, 0}, Point{len(grid[0]) - 1, len(grid) - 1}))
	fmt.Println("Part 2:", pathfind(grid, len(grid[0])*5, len(grid)*5, Point{0, 0}, Point{len(grid[0])*5 - 1, len(grid)*5 - 1}))
}

File: day15/src/main.rs
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashSet};
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

#[derive(Clone, Copy, Hash, Eq, PartialEq, PartialOrd, Ord, Debug)]
struct Point(usize, usize);
#[derive(Eq, PartialEq, Debug)]
struct Item(Point, usize, usize);

impl PartialOrd for Item {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Item {
    fn cmp(&self, Item(v, g, f): &Self) -> Ordering {
        let Item(sv, sg, sf) = self;
        f.cmp(sf).then_with(|| g.cmp(sg)).then_with(|| sv.cmp(v))
    }
}

fn abs(a: usize, b: usize) -> usize {
    if a < b {
        b - a
    } else {
        a - b
    }
}

fn manhattan(Point(ax, ay): Point, Point(bx, by): Point) -> usize {
    abs(ax, bx) + abs(ay, by)
}

fn neighbors(Point(x, y): Point) -> Vec<Point> {
    vec![
        Point(x.wrapping_sub(1), y),
        Point(x, y.wrapping_sub(1)),
        Point(x + 1, y),
        Point(x, y + 1),
    ]
}

fn in_bounds(Point(x, y): Point, w: usize, h: usize) -> bool {
    x < w && y < h
}

fn getval(grid: &Vec<Vec<usize>>, Point(x, y): Point, w: usize, h: usize) -> usize {
    (grid[y % h][x % w] + x / w + y / h - 1) % 9 + 1
}

fn pathfind(
    grid: &Vec<Vec<usize>>,
    vw: usize,
    vh: usize,
    start: Point,
    end: Point,
) -> Option<usize> {
    if grid.len() == 0 {
        return None;
    }
    let w = grid[0].len();
    let h = grid.len();
    let mut openset = BinaryHeap::new();
    openset.push(Item(start, 0, manhattan(start, end)));
    let mut closedset = HashSet::new();
    while let Some(Item(v, g, _)) = openset.pop() {
        if closedset.contains(&v) {
            continue;
        }
        closedset.insert(v);
        if v == end {
            return Some(g);
        }
        for i in neighbors(v) {
            if !in_bounds(i, vw, vh) || closedset.contains(&i) {
                continue;
            }
            let ng = g + getval(grid, i, w, h);
            openset.push(Item(i, ng, ng + manhattan(i, end)));
        }
    }
    None
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let grid = reader
        .lines()
        .flat_map(|i| i)
        .map(|line| {
            line.as_bytes()
                .into_iter()
                .map(|&i| (i - b'0') as usize)
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();
    println!(
        "Part 1: {}",
        pathfind(
            &grid,
            grid[0].len(),
            grid.len(),
            Point(0, 0),
            Point(grid[0].len() - 1, grid.len() - 1)
        )
        .ok_or("Fail part 1")?
    );
    println!(
        "Part 2: {}",
        pathfind(
            &grid,
            grid[0].len() * 5,
            grid.len() * 5,
            Point(0, 0),
            Point(grid[0].len() * 5 - 1, grid.len() * 5 - 1)
        )
        .ok_or("Fail part 2")?
    );
    Ok(())
}
